File: P_5/P5_PP_JOyAL.py
from collections import deque

class Ferris:
    def __init__(self, capacidad_maxima):
        self.capacidad_maxima = capacidad_maxima
        self.autos = deque()

    def insertar_auto(self, matricula):
        if len(self.autos) < self.capacidad_maxima:
            self.autos.append(matricula)
            return True
        else:
            return False

    def sacar_auto(self):
        raise NotImplementedError("Este método debe ser implementado en las clases hijas.")

    def imprimir_contenido(self):
        print("Contenido del ferri:")
        for matricula in self.autos:
            print(matricula)
        print()

class Portos(Ferris):
    def __init__(self, capacidad_maxima):
        super().__init__(capacidad_maxima)

    def sacar_auto(self):
        if len(self.autos) > 0:
            matricula = self.autos.pop()  # Utiliza pop() para sacar el último elemento de la pila
            return matricula
        else:
            return None


class Celebrian(Ferris):
    def __init__(self, capacidad_maxima):
        super().__init__(capacidad_maxima)

    def sacar_auto(self):
        if len(self.autos) > 0:
            matricula = self.autos.popleft()  # Utiliza popleft() para sacar el primer elemento de la cola
            return matricula
        else:
            return None

def main():
    print("xddddd")
    portos = Portos(10)
    celebrian = Celebrian(10)

    while True:
        matricula = input("Ingrese la matrícula del auto (o 'salir' para terminar): ")
        matricula = matricula.lower()
        if matricula == 'salir':
            break

        ferri = input("Ingrese el ferri al que irá el auto (P para Portos, C para Celebrian): ")
        ferri = ferri.lower()
        print(ferri)
        if ferri == 'p':
            if portos.insertar_auto(matricula):
                print("Auto ingresado en Portos.")
            else:
                print("Capacidad máxima de Portos alcanzada. No se pudo ingresar el auto.")
        elif ferri == 'c':
            if celebrian.insertar_auto(matricula):
                print("Auto ingresado en Celebrian.")
            else:
                print("Capacidad máxima de Celebrian alcanzada. No se pudo ingresar el auto.")
        else:
            print("Ferri inválido. Por favor, ingrese 'P' o 'C'.")

        portos.imprimir_contenido()
        celebrian.imprimir_contenido()

    print("Comenzando el viaje...")

    while True:
        matricula = portos.sacar_auto()
        if matricula is not None:
            print("Auto saliendo de Portos:", matricula)
        else:
            break

        matricula = celebrian.sacar_auto()
        if matricula is not None:
            print("Auto saliendo de Celebrian:", matricula)
        else:
            break

        portos.imprimir_contenido()
        celebrian.imprimir_contenido()

    print("Viaje completado.")
    
main()

File: P_5/P5_PP_JOyAL.java
import java.util.Stack;
import java.util.LinkedList;
import java.util.Scanner;

/*
    Practica 5
	Flores Lara Alberto y Saucillo Gonzalez Jesse Obed
	3BV2
 */

public class P5_PP_JOyAL {
    
    // Cola para almacenar los carros en el ferri Celebrian
    private static LinkedList<String> Celebrian = new LinkedList<String>();
    // Pila para almacenar los carros en el ferri Portos
    private static Stack<String> Portos = new Stack<String>();

    public static void main(String[] args) {
        int opc, ferri, bandera;
        int Capacidad = 10; // Capacidad máxima de cada ferri
        Scanner scanner = new Scanner(System.in);
        String Matricula;
        opc = 2; // Valor inicial para entrar al bucle do-while

        do {
            // Verificar si ambos ferris están llenos para iniciar la Simulacion
            if(Celebrian.size() == Capacidad && Portos.size() == Capacidad){
                System.out.println();
                opc = 2; 
                Simulacion();

                }else{
                    System.out.println("Bienvenido al Menu, Selecciona una opcion");
                    System.out.println("1. Ingresar un nuevo carro");
                    System.out.println("2. Adelantar la Simulacion de los Ferris");
                    opc = scanner.nextInt();
                    scanner.nextLine();
                    bandera = 0;

                    switch(opc){
                        case 1:
                            System.out.println("Ingrese la matrícula del Carro: ");
                            Matricula = scanner.nextLine();
                            // Seleccionar el ferri para subir el auto
                            do{
                                System.out.println("Seleccione en que Ferri desea subir el auto: ");
                                System.out.println("1. Portos");
                                System.out.println("2. Celebrian");
                                ferri = scanner.nextInt();
                                if (ferri == 1) {
                                    // Verificar si el ferri Portos está lleno
                                    if (Portos.size() == Capacidad) {
                                        System.out.println("El ferri Portos está lleno");
                                    } else {
                                        // Agregar la matrícula del carro al ferri Portos
                                        Portos.push(Matricula);
                                        System.out.println();
                                        imprimir();
                                        bandera = 1;
                                        break;
                                    }
                                } else if (ferri == 2) {
                                    // Verificar si el ferri Celebrian está lleno
                                    if (Celebrian.size() == Capacidad) {
                                        System.out.println("El ferri Celebrian está lleno");
                                    } else {
                                        // Agregar la matrícula del carro al ferri Celebrian
                                        Celebrian.offer(Matricula);
                                        System.out.println();
                                        imprimir();
                                        bandera = 1;
                                        break;
                                    }
                                } else {
                                    System.out.println("Selección inválida, intentelo de nuevo");
                                }
                            } while(bandera!=1);
                            break;

                        case 2:
                            System.out.println();
                            // Forzar Simulacion
                            Simulacion();
                            break;
                        
                        default:
                            System.out.println("Ingrese una opcion valida");
                    }
                }
            } while(opc!= 2);
        }
    
    // Método para simular el viaje de los ferris
    private static void Simulacion() {
        System.out.println("Simulacion de viaje del ferri Portos:");
        while (!Portos.isEmpty()) {
            System.out.println("Sale el Carro con matrícula " + Portos.pop());
            imprimir();
        }

        System.out.println();
    
        System.out.println("Simulacion de viaje del ferri Celebrian:");
        while (!Celebrian.isEmpty()) {
            System.out.println("Sale el Carro con matrícula " + Celebrian.poll());
            imprimir();
        }
    }
    
    // Método para imprimir el estado actual de los ferris
    private static void imprimir() {
        System.out.println("Ferri Portos: " + Portos);
        System.out.println("Ferri Celebrian: " + Celebrian);
        System.out.println();
    }

}
    


    


